dependencies/README.md

```diff
@@ -87,3 +89,80 @@ The tool allows users to choose the scope of entry point classes:
+  `classifier` is optional.
+- Edge: an (directed) edge is a dependency between Maven artifacts. A dependency from a Maven
+  artifact (_source_ of the dependency) to another artifact (_target_ of the dependency) is defined
+  in the `dependencies` element in pom.xml of a Maven artifact.
+  
+  A dependency has a boolean attribute `optional` and a string attribute `scope`,
+  among other properties listed in [POM Reference: Dependencies][1].
+
+  Self-loops or parallel edges are not possible.
+
+### Graph Construction
+
+Given a ordered list of Maven artifacts, a Maven dependency graph is constructed in the
+following manner:
+
+1. Start with a graph with nodes of the Maven artifacts in the list.
+   The nodes are called _initial nodes_.
+2. Pick up an unvisited node in a graph in breadth-first manner.
+3. By reading dependencies of the node, add new nodes corresponding to the target Maven artifacts,
+   identified by `groupId:artifactId[:classifier]:version`, if not present.
+4. Add edges from the source node to the target nodes of the Maven artifacts.
+3. Repeat the step 2-4, until all nodes are visited in the breadth-first traversal.
+
+A graph construction may _fail_ when there is a problem in constructing a graph (see below).
+
+### Edge Cases
+
+#### Cyclic Dependency
+
+Cyclic dependency may exist in a Maven dependency graph. It does not become a cause of a graph
+construction failure.
+
+#### Unavailable Artifact
+
+A Maven artifact may be unavailable through Maven repositories, making it impossible to complete
+a graph construction.
+
+An unavailability of Maven artifact is called _safe_ when the path from the initial nodes to
+the missing artifact contains `optional` or `scope: provided` dependency. An edge whose source
+artifact is missing but is safe, is skipped in a graph construction (step 4).
+
+When there is an unavailable Maven artifact and it is not safe, the graph construction fails.
+
+#### Unsatisfied Version Constraints
+
+A dependency element in pom.xml may have a [version range specification][2].
+Each of the specifications creates a version constraint.
+
+When there is a version constraint that cannot be satisfied during graph construction,
+the graph construction fails.
+
+### Class Path Generation through Maven Dependency Graph
+
+A class path (list of jar files of Maven artifacts) can be generated from a Maven dependency graph.
+A class path is built by picking up Maven artifacts by a breadth-first traversal,
+starting from the initial nodes of a Maven dependency graph.
+
+During the pick-up, duplicate artifacts identified by `groupId:artifactId[:classifier]:version`
+are discarded.
+
+When there are multiple versions of a Maven artifact identified by
+`groupId:artifactId[:classifier]` (without version), a version is picked up using one of following
+strategies:
+
+- **Maven dependency mediation strategy**: the version of the Maven artifact closest to the initial
```

Comment:
"closest" --> first version encountered during the breadth first traversal

for precision